### ISSUE TITLE
`list[T]` is incompatible with protocols that define `__getitem__()`

### ISSUE THREAD

**Issue by bb010g** [microsoft/pyright#3116]
### Describe the bug
`list[T]` is incompatible with protocols that define `__getitem__()`, even with matching overloads.

### To Reproduce
Typecheck the following file:
```py
from typing import Protocol, TypeVar


T = TypeVar('T')


class CopyableAsListProtocol(Protocol[T]):
    def copy(self) -> list[T]: ...


def passes_typecheck() -> CopyableAsListProtocol[bool]:
    foo = [False]
    return foo


class SliceableProtocol(Protocol[T]):
    def __getitem__(self, index: slice) -> list[T]: ...


def fails_typecheck() -> SliceableProtocol[bool]:
    bar = [False]
    return bar
```

`pyright test_3116.py`'s output:
```
No configuration file found.
No pyproject.toml file found.
stubPath /Users/dubanks/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
/Users/bb010g/test_3116.py
  /Users/bb010g/test_3116.py:22:12 - error: Expression of type "list[bool]" cannot be assigned to return type "SliceableProtocol[bool]"
    "list[bool]" is incompatible with protocol "SliceableProtocol[bool]"
      "__getitem__" is an incompatible type
        No overloaded function matches type "(index: slice) -> list[T@SliceableProtocol]"
          Type "(__i: SupportsIndex) -> bool" cannot be assigned to type "(index: slice) -> list[T@SliceableProtocol]"
            Parameter 1: type "slice" cannot be assigned to type "SupportsIndex"
            Position-only parameter mismatch; expected 1 but received 0
            Function return type "bool" is incompatible with type "list[T@SliceableProtocol]"
          Type "(__s: slice) -> list[bool]" cannot be assigned to type "(index: slice) -> list[T@SliceableProtocol]"
    ... (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

### Expected behavior
Pyright should widen `bar` from `list[bool]` to `IndexableAndSliceableProtocol[bool]`, like when it widens `foo` from `list[bool]` to `CopyableAsListProtocol[bool]`.

`mypy test_3116.py`'s output:
```
Success: no issues found in 1 source file
```

### VS Code extension or command-line
Pyright is at version 1.1.224.post1 from PyPI. mypy is at 0.931 from PyPI. These sample outputs are with the command line, but the behavior reproduces with the language server (controlled via Neovim's LSP client & configured via [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig)).

**Comment by erictraut** [comment]
Pyright is correct flagging this as an error. Your protocol `SliceableProtocol` indicates that the second parameter of `__getitem__` must accept a named argument `index`, but the built-in `list` class does not. Its second parameter to `__getitem__` is positional-only. You can eliminate the error by making renaming the parameter to `__index` within your protocol definition. Alternately, you can add a position-only parameter marker (`/`) after the `index` parameter.
